applicationset/generators/matrix.go

```diff
@@ -44,11 +44,11 @@ func (m *MatrixGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.App
 
 	res := []map[string]string{}
 
-	g0, err := m.getParams(appSetGenerator.Matrix.Generators[0], appSet)
+	g0, err := m.getParams(appSetGenerator.Matrix.Generators[0], appSet, []map[string]string{})
 	if err != nil {
 		return nil, err
 	}
-	g1, err := m.getParams(appSetGenerator.Matrix.Generators[1], appSet)
+	g1, err := m.getParams(appSetGenerator.Matrix.Generators[1], appSet, g0)
```

Comment:
Regarding your “I think it'll always be an array of size 1” comment from [the other thread](https://github.com/argoproj/argo-cd/pull/9080/files#r851585457), isn’t `g0` _all_ items from the first generator? And doesn’t it eventually become the `params` parameter to your new `interpolateGenerator` function? I’m unclear how the new example you added to `Generators-Matrix.md` can be satisfied unless `g1` is evaluated for each item in `g0`, instead of only for the last item in `g0` (which we seem to agree is what’s happening based on that loop in `interpolateGenerator`).

I suggest additional test case(s) in `matrix_test.go` to validate the complete results from a matrix generator using this functionality (specifically where the first generator yields multiple items).

I’m very interested in this PR for some of my own use cases and I’d try to help out but so far, I’ve been unsuccessful getting a local environment up and running.

Reply:
Hi @Lobstrosity,

So as to the example you were referring to, I'm not sure if you were able to check my *TestInterpolateGenerator* function. I had the following example which was very similar to the example I had in `Generators-Matrix.md` (actually it used more parameters than the example). As you can see in the below test case, the cluster-generator is using 3 different parameters that should be created by the git generator. 

<img width="896" alt="image" src="https://user-images.githubusercontent.com/29113963/164104082-5b9cf149-c756-4b32-9c84-cc8f7b1eb10c.png">

If `g1` does need to be evaluated for every item in `g0`, (by item, I'm assuming we're both referring to parameter generated for that particular generator), and that it currently only evaluates the last item, then wouldn't you agree that:
1) All the keys for this label-selector's values - Be evaluated with the last item in the params list (ie. `app3`)
2) All these test-cases be failing? 

<img width="935" alt="image" src="https://user-images.githubusercontent.com/29113963/164105635-f276a64c-e9b0-4664-8b32-e9f7dd99fcd1.png">

Reply:
Also though, I will try your above code snippet and see if I can observe a difference. 

```yaml
g0, err := m.getParams(appSetGenerator.Matrix.Generators[0], appSet, map[string]string{})

for _, a := range g0 {
  g1, err := m.getParams(appSetGenerator.Matrix.Generators[1], appSet, a)

  for _, b := range g1 {
    val, err := utils.CombineStringMaps(a, b)
    if err != nil {
      return nil, err
    }
    res = append(res, val)
  }
}
```

Reply:
Also @KojoRising m unsure the example in Generators-Matrix.md is workable with this code. i would be good to see a e2e test.